### PR TITLE
Restore Identity from Keyring Configuration

### LINKED ISSUE #471: Pipenv graph fails with  "cannot import name 'get_installed_distributions'"
```
  ☤  ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 1/1 — 00:00:04

To activate this project's virtualenv, run pipenv shell.
Alternatively, run a command inside the virtualenv with pipenv run.

ERROR:  Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/pipdeptree.py", line 17, in <module>
    from pip._internal import get_installed_distributions
ImportError: cannot import name 'get_installed_distributions'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/pipenv/vendor/pipdeptree.py", line 20, in <module>
    from pip import get_installed_distributions, FrozenRequirement
ImportError: cannot import name 'get_installed_distributions'

Exited with code 1
```

Looks like the Issue is die to incompatible version of `pip` and `pipenv` breaking `pipenv install`, `pipenv lock` (https://github.com/pypa/pipenv/issues/2924) and `pipenv graph` (see https://github.com/pypa/pip/issues/5154, https://github.com/pypa/pipenv/issues/2925) commands.

### PR DESCRIPTION
**Pull request by KPrasch**
Management and restoration of `CryptoPowers` from the filesystem; Allows uses to start/stop the *same* node when not in `--dev` mode.  Includes significant work on `NucypherKeyring` with assistance from @tuxxy 

- `NucypherKeyring` `.lock` and `.unlock` methods
- Derive `CryptoPowers` from `NucypherKeyring`
- Restore `HostingPower`, `EncryptingPower`, and `SigningPower` from disk
- Integrate `NucypherKeyring` with `NodeConfigurations`: Includes keyring in character pre-init logic (`NucypherKeyring.produce`) 
- Save and load node configurations from JSON
-  In-line Overridable `NodeConfiguration`s
- Account & Cryptographic keypair generation and storage via CLI
- CLI UX  improvements, color, and error handling options, Human Guidance
- Preserves "keyless"  temporary node with `--dev` mode

Fixes #471 #470 

*based on #465 *

### CODE REVIEW

File: nucypher/config/keyring.py [hunk 1]
```diff
@@ -285,6 +267,55 @@ def _generate_tls_keys(host: str, curve: EllipticCurve) -> Tuple[_EllipticCurveP
     return private_key, cert
 
 
+class _PrivateKeySerializer:
```
Comment (KPrasch) [inline nucypher/config/keyring.py]: @tuxxy

File: nucypher/config/keyring.py [hunk 2]
```diff
@@ -546,95 +588,82 @@ def generate(cls,
...
 
-            def __save_key(pk, encoding):
-                pem = pk.private_bytes(
-                    encoding=encoding,
+            def __serialize_pem(pk):
+                return pk.private_bytes(
+                    encoding=serialization.Encoding.PEM,
                     format=serialization.PrivateFormat.TraditionalOpenSSL,
-                    encryption_algorithm=serialization.NoEncryption()
+                    encryption_algorithm=serialization.BestAvailableEncryption(password=derived_key_material)
```
Comment (KPrasch) [inline nucypher/config/keyring.py]: @tuxxy

File: nucypher/config/keyring.py [hunk 3]
```diff
@@ -97,137 +134,207 @@ def _save_public_keyfile(keypath: str, key_data: bytes) -> str:
...
     ).derive(key_material)
-
-    return wrapping_key[:32]
+    return wrapping_key[:__WRAPPING_KEY_LENGTH]
 
 
-def _encrypt_umbral_key(wrapping_key: bytes, umbral_key: UmbralPrivateKey) -> dict:
+def _encrypt_umbral_key(wrapping_key: bytes,
+                        umbral_key: UmbralPrivateKey
+                        ) -> Dict[str, bytes]:
```
Comment (cygnusv) [inline nucypher/config/keyring.py]: What about using the `UmbralPrivateKey.to_bytes(password)` function?

Comment (KPrasch) [reply nucypher/config/keyring.py]: Yes - Good Idea - Made this change in #480; Although, now derivation time for new accounts is > 5 seconds.

Comment (cygnusv) [reply nucypher/config/keyring.py]: yeah, you can control that with the `_scrypt_cost` argument. Default value is 20. Anything between 15 and 20 should be fine. Note that the time grows exponentially wrt to this argument. so if you want to make it 4 times faster, a cost of 18 should be OK.

Comment (tuxxy) [reply nucypher/config/keyring.py]: Standard cost according to the RFC is 20 for symmetric key derivation.

**Pull request merged**